modules/renter/downloadbuffer.go

```diff
@@ -0,0 +1,172 @@
+package renter
+
+// Downloads can be written directly to a file, can be written to an http
+// stream, or can be written to an in-memory buffer. The core download loop only
+// has the concept of writing using WriteAt, so to support writing to a stream
+// or to an in-memory buffer, we need to wrap the function with something that
+// will transform the WriteAt call into an in-order stream or otherwise write it
+// to the right place.
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+// downloadDestination is a wrapper for the different types of writing that we
+// can do when reovering and writing the logical data of a file. The wrapper
+// needs to convert the various write-at calls into writes that make sense to
+// the underlying file, buffer, or stream.
+//
+// For example, if the underlying object is a file, the WriteAt call is just a
+// passthrough function. But if the underlying object is a stream, WriteAt may
+// block while it waits for previous data to be written.
+type downloadDestination interface {
+	Close() error
+	WriteAt(data []byte, offset int64) (int, error)
+}
+
+// downloadDestinationBuffer writes logical chunk data to an in-memory buffer.
+// This buffer is primarily used when performing repairs on uploads.
+type downloadDestinationBuffer []byte
+
+// Close implements Close for the downloadDestination interface.
+func (dw downloadDestinationBuffer) Close() error {
+	return nil
+}
+
+// WriteAt writes the provided data to the downloadDestinationBuffer.
+func (dw downloadDestinationBuffer) WriteAt(data []byte, offset int64) (int, error) {
+	if len(data)+int(offset) > len(dw) || offset < 0 {
+		return 0, errors.New("write at specified offset exceeds buffer size")
+	}
+	i := copy(dw[offset:], data)
+	return i, nil
+}
+
+// downloadDestinationWriter is a downloadDestination that writes to an
+// underlying data stream. The data stream is expecting sequential data while
+// the download chunks will be written in an aribtrary order using calls to
+// WriteAt. We need to block the calls to WriteAt until all prior data has been
+// written.
+//
+// TODO: There is no timeout protection here. If there's some misalignment of
+// data, we'll never know it'll just hang forever.
+type downloadDestinationWriter struct {
```

Comment:
I know I explained it in a comment somewhere, will have to explain it here as well. The download code ensures that it allocates memory for all earlier chunks before it begins work on any later chunks. So even if a later chunk finishes first, we know that the earlier chunk at the very least has had memory allocated, so it will be finishing up soon.

But... now that I think about it I don't think we have the `Close()` function implemented correctly here. It needs to unblock all of the other threads and have them fail.

Reply:
so there is a potential deadlock here

Reply:
Ahh yeah I found the Note. Might be a bit tricky to find all the other chunks and make them fail or remove them from the heap after the first one failed though.